OpenRA.SourceGenerators/Sync/SyncHashCodeSourceGenerator.cs

```diff
@@ -0,0 +1,106 @@
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace OpenRA.SourceGenerators.Sync
+{
+	// https://andrewlock.net/exploring-dotnet-6-part-9-source-generator-updates-incremental-generators/
+	[Generator]
+	public class SyncHashCodeSourceGenerator : IIncrementalGenerator
+	{
+		public void Initialize(IncrementalGeneratorInitializationContext context)
+		{
+			var classDeclarationsProvider = context.SyntaxProvider
+				.CreateSyntaxProvider(
+					predicate: static (syntaxNode, _) => syntaxNode.IsPartialClass(),
+					transform: static (generatorContext, _) => GetClassDeclarationSymbol(generatorContext))
+				.Where(x => x != null)
+				.Collect();
+
+			var compilationAndClasses = context.CompilationProvider.Combine(classDeclarationsProvider);
+
+			context.RegisterSourceOutput(compilationAndClasses,
+				static (context, source) => GenerateCode(source.Left, source.Right, context));
+		}
+
+		static ClassDeclarationSyntax GetClassDeclarationSymbol(GeneratorSyntaxContext context)
+		{
+			if (context.SemanticModel.GetDeclaredSymbol(context.Node) is INamedTypeSymbol classSymbol && classSymbol.ImplementsISync())
+				return context.Node as ClassDeclarationSyntax;
+
+			return null;
+		}
+
+		static void GenerateCode(Compilation compilation, ImmutableArray<ClassDeclarationSyntax> classes, SourceProductionContext context)
+		{
+			if (classes.IsDefaultOrEmpty)
+				return;
+
+			foreach (var classDeclaration in classes.Distinct())
+			{
+				var (filename, content) = ProcessType(classDeclaration, compilation);
+				if (content != null)
+					context.AddSource(filename, content);
+			}
+		}
+
+		static (string FileName, string Content) ProcessType(ClassDeclarationSyntax classDeclaration, Compilation compilation)
+		{
+			if (classDeclaration.Parent is not NamespaceDeclarationSyntax namespaceDeclaration)
+				return (null, null);
+
+			var semanticModel = compilation.GetSemanticModel(classDeclaration.SyntaxTree);
+			var typeSymbol = semanticModel.GetDeclaredSymbol(classDeclaration) as ITypeSymbol;
+			var syncedClassMembers = typeSymbol.GetAllMembers(x => x.HasSyncAttribute()).ToArray();
+			var hashCodeStrings = new List<string>();
+			var syncsTargets = false;
+			foreach (var member in syncedClassMembers)
+			{
+				var isTarget = false;
+				if (member is IFieldSymbol field)
+					hashCodeStrings.Add(SyncHelpers.GetHashCodeString(field.Type.Name, member.Name, out isTarget));
+
+				if (member is IPropertySymbol property)
+					hashCodeStrings.Add(SyncHelpers.GetHashCodeString(property.Type.Name, member.Name, out isTarget));
+
+				syncsTargets |= isTarget;
+			}
+
+			var isSealed = classDeclaration.IsSealed();
+
+			return GenerateClass(namespaceDeclaration.Name, classDeclaration.Identifier.Text, hashCodeStrings, isSealed, syncsTargets);
```

Comment:
If my `partial` class already implements `ISync` directly in the type (and not from a base type) then we shouldn't generate code for this, or else C# complains that we're trying to implement the interface twice.